examples/custom_clusterer.py

```diff
@@ -0,0 +1,219 @@
+    return epsilon_density_clusterer(epsilon, k, data)
+
+# clustering parameters
+epsilon = 0.1
+k = 4
+
+# %%
+# we can then visualize to see if this clusterer acts as we would expect.
+
+# dataset parameters
+n_points = 4000 
+noise_level = 0.1  # Noise level for the radius
+
+# generate angles uniformly between 0 and 2*pi
+angles = np.linspace(0, 2 * np.pi, n_points)
+
+# generate radii close to 1 with some noise
+radii = 1 + noise_level * np.random.randn(n_points)
+
+# convert polar coordinates to Cartesian coordinates
+x = radii * np.cos(angles)
+y = radii * np.sin(angles)
+
+
+# stack x and y into a 2D array for clustering
+data = np.column_stack((x, y))
+
+clusters = list(epsilon_density_clusterer(epsilon=epsilon, k=k, data=data))
+
+# create an array for cluster labels
+cluster_labels = np.full(data.shape[0], -1)  # use default label for noise points
+for cluster_id, indices in enumerate(clusters):
+    cluster_labels[indices] = cluster_id
+
+# plotting
+plt.figure(figsize=(10, 6))
+scatter = plt.scatter(
+    data[:, 0], 
+    data[:, 1], 
+    c=cluster_labels, 
+    cmap='viridis', 
+    edgecolor='k', 
+    s=30, 
+    alpha=.5
+    )
+plt.colorbar(scatter, label='Cluster ID')
+plt.title('Density Clustering')
+plt.xlabel('x')
+plt.ylabel('y')
+plt.axis('equal')
+plt.grid('True')
+plt.show()
+
+
+
+
+# %%
+# Using with mapper
+# =================
+#
+# Now we have a clusterer compatible with zen-mapper.
+
+cover_scheme = Width_Balanced_Cover(
+    n_elements=5, 
+    percent_overlap=0.25
+    )
+cover = cover_scheme(data)
+projection = data[:,0]
+
+result = mapper(
+    data=data,
+    projection=projection,
+    cover_scheme=cover_scheme,
+    clusterer=clusterer,
+    dim=1
+)
+
+graph = to_networkx(result.nerve)
+
+# plot the mapper graph
+nx.draw_kamada_kawai(graph)
+plt.show()
+
+
+
+# %%
+# Coloring the nodes
+# ==================
+
+density_counts = neighbor_counts(epsilon=epsilon, data=data)
+
+node_densities = {}
+
+for node_id, indices in enumerate(result.nodes):
+    # calculate the average density for the current cluster
+    node_densities[node_id] = np.mean(density_counts[indices])
+
+# create a color map based on node densities
+node_colors = [node_densities[node] for node in graph.nodes]
+
+# Define the color map for the density values
+cmap = plt.cm.viridis
+norm = mpl.colors.Normalize(
+    vmin=min(node_colors), 
+    vmax=max(node_colors)
+    )
+sm = mpl.cm.ScalarMappable(cmap=cmap, norm=norm)
+sm.set_array([])  # ScalarMappable needs an array, though it's not used here.
+
+# set up figure and axis
+fig, ax = plt.subplots(figsize=(10, 6))
+
+# plot the mapper graph with nodes colored by density
+pos = nx.kamada_kawai_layout(graph)  # use a layout for better visualization
+nx.draw(
+    graph,
+    pos,
+    node_color=node_colors,
+    cmap=cmap,
+    node_size=100,
+    ax=ax
+)
+
+# add the color bar for average density, specifying the axis
+cbar = fig.colorbar(sm, ax=ax, label='Average Density')
```

Comment:
You can avoid having to manually create this color map by calling `nx.draw_networkx_nodes` instead
```python
# plot the mapper graph with nodes colored by density
pos = nx.kamada_kawai_layout(graph)
sm = nx.draw_networkx_nodes(
    graph,
    pos,
    node_color=node_colors,
    node_size=100,
    ax=ax,
)
nx.draw_networkx_edges(graph, pos)

# add the color bar for average density, specifying the axis
cbar = fig.colorbar(sm, ax=ax, label="Average Density")
```